File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.cpp
#include <string>
#include <vector>
#include <queue>
#include <climits>

using namespace std;

int dx[4] = {1, 0, -1, 0};
int dy[4] = {0, 1, 0, -1};
 
int solution(vector<vector<int>> rectangle, int characterX, int characterY, int itemX, int itemY) {
    int MAX = 102;
    vector<vector<int>> board(MAX, vector<int>(MAX, 0));
    
    for (auto& r : rectangle) {
        int x1 = r[0] * 2, y1 = r[1] * 2, x2 = r[2] * 2, y2 = r[3] * 2;
        for (int i = x1; i <= x2; i++) {
            for (int j = y1; j <= y2; j++) board[i][j] = 1;
        }
    }
    
    for (auto& r : rectangle) {
        int x1 = r[0] * 2, y1 = r[1] * 2, x2 = r[2] * 2, y2 = r[3] * 2;
        for (int i = x1 + 1; i < x2; i++) {
            for (int j = y1 + 1; j < y2; j++) board[i][j] = 0;
        }
    }
    
    int sx = characterX * 2, sy = characterY * 2;
    int tx = itemX * 2, ty = itemY * 2;
    
    queue<pair<int, int>> q;
    q.push({sx, sy});
    
    vector<vector<int>> dist(MAX, vector<int>(MAX, -1));
    dist[sx][sy] = 0;
    
    while(!q.empty()) {
        auto [x, y] = q.front();
        q.pop();
        
        if (x == tx && y == ty) break;
        
        for (int dir = 0; dir < 4; dir++) {
            int nx = x + dx[dir];
            int ny = y + dy[dir];
            if (nx < 0 || nx >= MAX || ny < 0 || ny >= MAX) continue;
            if (board[nx][ny] == 0) continue;
            if (dist[nx][ny] != -1) continue;
            q.push({nx, ny});
            dist[nx][ny] = dist[x][y] + 1;
        }
    }
    
    return dist[tx][ty] / 2;
}

File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.java
import java.util.*;

class Solution {
    public int solution(int[][] rectangle, int characterX, int characterY, int itemX, int itemY) {
        int[][] board = new int[102][102];
        for (int[] rect : rectangle) {
            int x1 = rect[0]*2, y1 = rect[1]*2, x2 = rect[2]*2, y2 = rect[3]*2;
            for (int i = x1; i <= x2; i++) {
                for (int j = y1; j <= y2; j++) {
                    board[i][j] = 1;
                }
            }
        }
        
        for (int[] rect : rectangle) {
            int x1 = rect[0]*2, y1 = rect[1]*2, x2 = rect[2]*2, y2 = rect[3]*2;
            for (int i = x1+1; i < x2; i++) {
                for (int j = y1+1; j < y2; j++) {
                    board[i][j] = 0;
                }
            }
        }
        
        int answer = bfs(board, characterX*2, characterY*2, itemX*2, itemY*2);
        
        return answer / 2;
    }
    
    int[] dx = {1, 0, -1, 0};
    int[] dy = {0, 1, 0, -1};
    
    private int bfs(int[][] board, int startx, int starty, int endx, int endy) {
        Queue<int[]> q = new LinkedList<>();
        boolean[][] visited = new boolean[102][102];
        q.offer(new int[]{startx, starty});
        visited[startx][starty] = true;
        
        while(!q.isEmpty()) {
            int[] cur = q.poll();
            int x = cur[0];
            int y = cur[1];
            
            if (x == endx && y == endy) return board[endx][endy];
            
            for (int dir = 0; dir < 4; dir++) {
                int nx = x + dx[dir];
                int ny = y + dy[dir];
                
                if (nx < 0 || nx >= 102 || ny < 0 || ny >= 102) continue;
                if (visited[nx][ny]) continue;
                if (board[nx][ny] == 0) continue;
                
                visited[nx][ny] = true;
                q.offer(new int[]{nx, ny});
                board[nx][ny] = board[x][y] + 1;
            }
        }
        return -1;
    }
}